File: ios/TacticalMaps/Map/WeatherService.swift
import Foundation
import CoreLocation

/// Current conditions for a coordinate, pulled from Open-Meteo (same as
/// elevation lookups, no API key needed). Units: °C, m/s, metres.
struct WeatherReading: Equatable {
    let temperatureC: Double?
    let windSpeedMs: Double?
    let windGustsMs: Double?
    let visibilityM: Double?
    let weatherCode: Int?
}

/// UAV flight risk from a WeatherReading vs configurable thresholds.
/// Takes the worst-of component risk, so one bad factor (e.g. high gusts)
/// flags the whole thing.
enum UAVRisk: Int, Comparable {
    case safe = 0, caution = 1, danger = 2
    static func < (l: UAVRisk, r: UAVRisk) -> Bool { l.rawValue < r.rawValue }

    var label: String {
        switch self {
        case .safe:    return "Safe to fly"
        case .caution: return "Marginal — caution"
        case .danger:  return "Do not fly"
        }
    }
}

/// Default thresholds for small/consumer drones. Prob make these configurable later.
struct UAVThresholds {
    var windCautionMs = 7.0,  windDangerMs = 10.0
    var gustCautionMs = 8.0,  gustDangerMs = 12.0
    var visCautionM   = 5000.0, visDangerM = 1500.0
    var tempLowDangerC = -10.0, tempLowCautionC = 0.0
    var tempHighCautionC = 40.0, tempHighDangerC = 45.0
    static let `default` = UAVThresholds()
}

enum UAVAssessment {
    /// Worst-of risk across all components. Missing values just get skipped.
    static func risk(for r: WeatherReading, _ t: UAVThresholds = .default) -> UAVRisk {
        var level: UAVRisk = .safe
        func bump(_ x: UAVRisk) { if x > level { level = x } }

        if let w = r.windSpeedMs {
            if w >= t.windDangerMs { bump(.danger) } else if w >= t.windCautionMs { bump(.caution) }
        }
        if let g = r.windGustsMs {
            if g >= t.gustDangerMs { bump(.danger) } else if g >= t.gustCautionMs { bump(.caution) }
        }
        if let v = r.visibilityM {
            if v <= t.visDangerM { bump(.danger) } else if v <= t.visCautionM { bump(.caution) }
        }
        if let temp = r.temperatureC {
            if temp <= t.tempLowDangerC || temp >= t.tempHighDangerC { bump(.danger) }
            else if temp <= t.tempLowCautionC || temp >= t.tempHighCautionC { bump(.caution) }
        }
        return level
    }
}

/// Fetch current conditions from Open-Meteo forecast endpoint.
actor WeatherService {

    private struct Response: Decodable {
        struct Current: Decodable {
            let time: String?
            let temperature_2m: Double?
            let wind_speed_10m: Double?
            let wind_gusts_10m: Double?
            let weather_code: Int?
        }
        struct Hourly: Decodable {
            let time: [String]?
            let visibility: [Double]?
        }
        let current: Current?
        let hourly: Hourly?
    }

    func reading(for coordinate: CLLocationCoordinate2D) async -> WeatherReading? {
        // OPSEC: this sends the coordinate to Open-Meteo, so bail out
        // unless user opted into online lookups.
        guard OpsecSettings.shared.onlineLookups else { return nil }
        if coordinate.latitude == 0 && coordinate.longitude == 0 { return nil }
        // Coarsen to ~110 m (3 dp) so the exact position isn't disclosed.
        let coordinate = CLLocationCoordinate2D(
            latitude: (coordinate.latitude * 1000).rounded() / 1000,
            longitude: (coordinate.longitude * 1000).rounded() / 1000)

        var components = URLComponents(string: "https://api.open-meteo.com/v1/forecast")
        components?.queryItems = [
            URLQueryItem(name: "latitude",  value: String(coordinate.latitude)),
            URLQueryItem(name: "longitude", value: String(coordinate.longitude)),
            URLQueryItem(name: "current",   value: "temperature_2m,wind_speed_10m,wind_gusts_10m,weather_code"),
            URLQueryItem(name: "hourly",    value: "visibility"),
            URLQueryItem(name: "wind_speed_unit", value: "ms"),
            URLQueryItem(name: "forecast_days", value: "1"),
            URLQueryItem(name: "timezone", value: "auto")
        ]
        guard let url = components?.url else { return nil }

        var request = URLRequest(url: url)
        request.timeoutInterval = 8
        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            let decoded = try JSONDecoder().decode(Response.self, from: data)
            guard let c = decoded.current else { return nil }
            return WeatherReading(
                temperatureC: c.temperature_2m,
                windSpeedMs:  c.wind_speed_10m,
                windGustsMs:  c.wind_gusts_10m,
                visibilityM:  visibilityNow(decoded, currentTime: c.time),
                weatherCode:  c.weather_code
            )
        } catch {
            return nil
        }
    }

    /// Grab hourly visibility matching current hour, falls back
    /// to first entry if we can't find a match.
    private func visibilityNow(_ r: Response, currentTime: String?) -> Double? {
        guard let times = r.hourly?.time, let vis = r.hourly?.visibility, !vis.isEmpty else { return nil }
        if let currentTime, let idx = times.firstIndex(of: currentTime), idx < vis.count {
            return vis[idx]
        }
        return vis.first
    }
}

File: android/app/src/main/java/com/tacmap/map/WeatherService.kt
package com.tacmap.map

import com.tacmap.settings.OpsecSettings
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.math.roundToInt

/** Current-conditions reading from Open-Meteo (same provider as elevation).
 *  Units: °C, m/s, metres. */
data class WeatherReading(
    val temperatureC: Double?,
    val windSpeedMs: Double?,
    val windGustsMs: Double?,
    val visibilityM: Double?,
    val weatherCode: Int?
)

/** UAV flight-safety risk, overall = worst of all components. */
enum class UAVRisk(val label: String) {
    SAFE("Safe to fly"),
    CAUTION("Marginal — caution"),
    DANGER("Do not fly")
}

/** Default UAV thresholds, small/consumer drone oriented. Tunable later. */
data class UAVThresholds(
    val windCautionMs: Double = 7.0, val windDangerMs: Double = 10.0,
    val gustCautionMs: Double = 8.0, val gustDangerMs: Double = 12.0,
    val visCautionM: Double = 5000.0, val visDangerM: Double = 1500.0,
    val tempLowDangerC: Double = -10.0, val tempLowCautionC: Double = 0.0,
    val tempHighCautionC: Double = 40.0, val tempHighDangerC: Double = 45.0
)

object UAVAssessment {
    /** Worst-of-components risk. Missing values don't raise risk. */
    fun risk(r: WeatherReading, t: UAVThresholds = UAVThresholds()): UAVRisk {
        var level = UAVRisk.SAFE
        fun bump(x: UAVRisk) { if (x.ordinal > level.ordinal) level = x }

        r.windSpeedMs?.let { if (it >= t.windDangerMs) bump(UAVRisk.DANGER) else if (it >= t.windCautionMs) bump(UAVRisk.CAUTION) }
        r.windGustsMs?.let { if (it >= t.gustDangerMs) bump(UAVRisk.DANGER) else if (it >= t.gustCautionMs) bump(UAVRisk.CAUTION) }
        r.visibilityM?.let { if (it <= t.visDangerM) bump(UAVRisk.DANGER) else if (it <= t.visCautionM) bump(UAVRisk.CAUTION) }
        r.temperatureC?.let {
            if (it <= t.tempLowDangerC || it >= t.tempHighDangerC) bump(UAVRisk.DANGER)
            else if (it <= t.tempLowCautionC || it >= t.tempHighCautionC) bump(UAVRisk.CAUTION)
        }
        return level
    }
}

/** Fetches current conditions from Open-Meteo's forecast endpoint. */
class WeatherService {

    @Serializable
    private data class Response(
        val current: Current? = null,
        val hourly: Hourly? = null
    ) {
        @Serializable
        data class Current(
            val time: String? = null,
            val temperature_2m: Double? = null,
            val wind_speed_10m: Double? = null,
            val wind_gusts_10m: Double? = null,
            val weather_code: Int? = null
        )
        @Serializable
        data class Hourly(
            val time: List<String> = emptyList(),
            val visibility: List<Double> = emptyList()
        )
    }

    private val json = Json { ignoreUnknownKeys = true }

    suspend fun reading(lat: Double, lng: Double): WeatherReading? {
        // OPSEC: weather lookup transmits coord to Open-Meteo, only proceed
        // if user opted in. Without this the dialog leaked map-centre even
        // with online lookups off.
        if (OpsecSettings.shared?.onlineLookups?.value != true) return null
        if (lat == 0.0 && lng == 0.0) return null
        // Coarsen to ~110m so exact position isn't disclosed.
        val cLat = (lat * 1000).roundToInt() / 1000.0
        val cLng = (lng * 1000).roundToInt() / 1000.0
        val url = "https://api.open-meteo.com/v1/forecast" +
            "?latitude=$cLat&longitude=$cLng" +
            "&current=temperature_2m,wind_speed_10m,wind_gusts_10m,weather_code" +
            "&hourly=visibility&wind_speed_unit=ms&forecast_days=1&timezone=auto"

        val body = boundedHttpsGet(url, MAX_RESPONSE_BYTES) ?: return null

        val decoded = runCatching { json.decodeFromString<Response>(body) }.getOrNull() ?: return null
        val c = decoded.current ?: return null
        return WeatherReading(
            temperatureC = c.temperature_2m,
            windSpeedMs = c.wind_speed_10m,
            windGustsMs = c.wind_gusts_10m,
            visibilityM = visibilityNow(decoded, c.time),
            weatherCode = c.weather_code
        )
    }

    private fun visibilityNow(r: Response, currentTime: String?): Double? {
        val times = r.hourly?.time ?: return null
        val vis = r.hourly.visibility
        if (vis.isEmpty()) return null
        val idx = currentTime?.let { times.indexOf(it) } ?: -1
        return if (idx in vis.indices) vis[idx] else vis.firstOrNull()
    }

    private companion object { const val MAX_RESPONSE_BYTES = 1024 * 1024 }
}
